unittests/producer_schedule_tests.cpp

```diff
@@ -503,4 +503,224 @@ BOOST_AUTO_TEST_CASE( producer_watermark_test ) try {
+   BOOST_REQUIRE_EQUAL( control->pending_block_producer(), N(alice) );
+
+   // produce with backup key by listing that in block signing keys
+   block_signing_private_keys.clear();
+   block_signing_private_keys.emplace(get_public_key(N(alice), "bs2"), get_private_key(N(alice), "bs2"));
+   block_signing_private_keys.emplace(get_public_key(N(bob),   "bs2"), get_private_key(N(bob),   "bs2"));
+
+   produce_blocks(config::producer_repetitions, false);
+
+   // check that its bobs turn
+   BOOST_REQUIRE_EQUAL( control->head_block_producer(),    N(alice) );
+   BOOST_REQUIRE_EQUAL( control->pending_block_producer(), N(bob) );
+
+   produce_blocks(config::producer_repetitions, false);
+
+   BOOST_REQUIRE_EQUAL( validate(), true );
+} FC_LOG_AND_RETHROW()
+
+BOOST_FIXTURE_TEST_CASE( producer_m_of_n_test, TESTER ) try {
+   create_accounts( {N(alice),N(bob)} );
+   while (control->head_block_num() < 3) {
+      produce_block();
+   }
+
+   auto compare_schedules = [&]( const vector<producer_authority>& a, const producer_authority_schedule& b ) {
+      return std::equal( a.begin(), a.end(), b.producers.begin(), b.producers.end() );
+   };
+
+   vector<producer_authority> sch1 = {
+                                 producer_authority{N(alice), block_signing_authority_v0{2, {{get_public_key(N(alice), "bs1"), 1}, {get_public_key(N(alice), "bs2"), 1}}}},
+                                 producer_authority{N(bob),   block_signing_authority_v0{2, {{get_public_key(N(bob),   "bs1"), 1}, {get_public_key(N(bob),   "bs2"), 1}}}}
+                               };
+
+   auto res = set_producer_schedule( sch1 );
+   block_signing_private_keys.emplace(get_public_key(N(alice), "bs1"), get_private_key(N(alice), "bs1"));
+   block_signing_private_keys.emplace(get_public_key(N(alice), "bs2"), get_private_key(N(alice), "bs2"));
+   block_signing_private_keys.emplace(get_public_key(N(bob),   "bs1"), get_private_key(N(bob),   "bs1"));
+   block_signing_private_keys.emplace(get_public_key(N(bob),   "bs2"), get_private_key(N(bob),   "bs2"));
+
+   //wdump((fc::json::to_pretty_string(res)));
+   wlog("set producer schedule to [alice,bob]");
+   BOOST_REQUIRE_EQUAL( true, control->proposed_producers().valid() );
+   BOOST_CHECK_EQUAL( true, compare_schedules( sch1, *control->proposed_producers() ) );
+   BOOST_CHECK_EQUAL( control->pending_producers().version, 0u );
+   produce_block(); // Starts new block which promotes the proposed schedule to pending
+   BOOST_CHECK_EQUAL( control->pending_producers().version, 1u );
+   BOOST_CHECK_EQUAL( true, compare_schedules( sch1, control->pending_producers() ) );
+   BOOST_CHECK_EQUAL( control->active_producers().version, 0u );
+   produce_block();
+   produce_block(); // Starts new block which promotes the pending schedule to active
+   BOOST_CHECK_EQUAL( control->active_producers().version, 1u );
+   BOOST_CHECK_EQUAL( true, compare_schedules( sch1, control->active_producers() ) );
+   produce_blocks(18);
```

Comment:
See earlier comment about `BOOST_REQUIRE( produce_empty_blocks_until( *this, N(bob), N(alice), 36 ) );`.